lib/Widgets/MessageDialog.vala

```diff
@@ -0,0 +1,261 @@
+     *
+     * Most of the times, you will only want to modify the {@link Granite.MessageDialog.secondary_text} string, 
+     * this is available to set additional properites like {@link Gtk.Label.use_markup} if you wish to do so.
+     */
+    public Gtk.Label secondary_label { get; construct; }
+
+    /**
+     * The {@link Gtk.ButtonsType} value to display a set of buttons
+     * in the dialog.
+     * 
+     * By design, some actions are not acceptable and such action values will not be added to the dialog, these include:
+     *
+     *  * {@link Gtk.ButtonsType.OK}
+     *  * {@link Gtk.ButtonsType.YES_NO}
+     *  * {@link Gtk.ButtonsType.OK_CANCEL}
+     *
+     * If you wish to provide more specific actions for your dialog
+     * pass a {@link Gtk.ButtonsType.NONE} to {@link Granite.MessageDialog.MessageDialog} and manually
+     * add those actions with {@link Gtk.Dialog.add_buttons} or {@link Gtk.Dialog.add_action_widget}.
+     */
+    public Gtk.ButtonsType buttons { 
+        construct {
+            switch (value) {
+                case Gtk.ButtonsType.NONE:
+                    break;
+                case Gtk.ButtonsType.CLOSE:
+                    add_button (_("_Close"), Gtk.ResponseType.CLOSE);
+                    break;
+                case Gtk.ButtonsType.CANCEL:
+                    add_button (_("_Cancel"), Gtk.ResponseType.CANCEL);
+                    break;
+                case Gtk.ButtonsType.OK:
+                case Gtk.ButtonsType.YES_NO:
+                case Gtk.ButtonsType.OK_CANCEL:
+                    warning ("Unsupported GtkButtonsType value");
+                    break;
+                default:
+                    warning ("Unknown GtkButtonsType value");
+                    break;
+            }
+        }
+    }
+
+    /**
+     * The custom area to add custom widgets.
+     * 
+     * This bin can be used to add any custom widget to the message area such as a {@link Gtk.ComboBox} or {@link Gtk.CheckButton}.
+     * Note that after adding such widget you will need to call {@link Gtk.Widget.show} or {@link Gtk.Widget.show_all} on the widget itself for it to appear in the dialog.
+     * 
+     * If you want to add multiple widgets to this area, create a new container such as {@link Gtk.Grid} or {@link Gtk.Box} and then add it to the custom bin.
+     * 
+     * When adding a custom widget to the custom bin, the {@link Granite.MessageDialog.secondary_label}'s bottom margin will be expanded automatically
+     * to compensate for the additional widget in the dialog.
+     * Removing the previously added widget will remove the bottom margin.
+     * 
+     * If you don't want to have any margin between your custom widget and the {@link Granite.MessageDialog.secondary_label}, simply add your custom widget
+     * and then set the {@link Gtk.Label.margin_bottom} of {@link Granite.MessageDialog.secondary_label} to 0.
+     */
+    public Gtk.Bin custom_bin { get; construct; }
```

Comment:
Using `Gtk.Bin` here sounds like a pain in the ass. Why not just use Grid and save having to create a grid?

Reply:
@danrabbit main problem is, having a Grid here disables us from having an ability to listen when a widget is added. E.g: calling `attach` on a Grid won't give us any signal that the widget was actually added. If we want a grid, the API user will have to set the margin on the secondary label manually.

Reply:
Ah okay that makes sense :+1: